juju/juju
# [dev] Support endpoints in port-related hook tools

## Description of change

This PR builds on top of the changes introduced by #11877 (the OpenedMachinePorts API call) to enable the port-related hook tools (open-ports, close-ports, opened-ports) to work with endpoints. 

The **default** mode of operation for the aforementioned tools has been preserved as-is for backwards compatibility reasons:
- `open/close-ports` (unless explicitly instructed) assume that open/close requests always apply to **all** application endpoints. This matches the pre-2.9 juju behavior.
- `opened-ports` will list the _unique_ set of open port ranges without any endpoint details (even if opened for a specific endpoint). This ensures that we don't break existing charms (and/or frameworks) that might be attempting to parse the command output.

Newer charms that want to leverage this new feature (they can use version sniffing to see if it is supported) can provide the `--endpoints` flag to the tools. More specifically:
- the `open/close-ports` tools allow charms to target the open/close request to a comma delimited list of endpoints. 
- the `opened-ports` tool will group opened ports by endpoint and include the list the endpoints (or `*` if the range is open across **all** endpoints) in its output.

## Implementation details

The PR includes a few important refactoring changes:
- the duplicated code for parsing and validating port ranges (and tests) has been removed from the hook tools; instead, we now use the (thoroughly tested) `core/network.ParsePortRange` to do the parsing/validation steps.
- the port range logic in the uniter's context has been rewritten from scratch so that it works with the various edge cases you might bump into when working with endpoints (e.g. a unit opens the same range for different endpoints). The new version is wrapped as a neat helper type which does not suffer from the pass-map-by-value (and then modify) issues of the previous implementation.
- several methods in the path from the hook tools to the uniter context operations that used to take a [from, to, protocol] tuple as argument have been changed to accept a port range.

Finally, the help docs for these hook tools have been rewritten so they match the style and layout of other hook tools' help docs.

## QA steps

*Please replace with how we can verify that the change works?*

```console
$ juju bootstrap lxd test --no-gui

$ juju deploy percona-cluster

# Mysql opens 3306 for all endpoints
$ juju run --unit percona-cluster/0 "opened-ports"
3306/tcp

# Check the detailed port listing
$ juju run --unit percona-cluster/0 "opened-ports --endpoints"
3306/tcp (*)

# Close 3306 for the db-admin and shared-db endpoints
# Juju should carve out the port from these endpoints and
# retain it for the rest
$ juju run --unit percona-cluster/0 "close-port 3306/tcp --endpoints db-admin,shared-db"

$ juju run --unit percona-cluster/0 "opened-ports --endpoints"
3306/tcp (access, cluster, db, ha, master, nrpe-external-master, slave)

# Open 1337 for the db-admin endpoint
$ juju run --unit percona-cluster/0 "open-port 1337/udp --endpoints db-admin"

$ juju run --unit percona-cluster/0 "opened-ports --endpoints"
3306/tcp (access, cluster, db, ha, master, nrpe-external-master, slave)
1337/udp (db-admin)

# Open 3306 for *all* endpoints
$ juju run --unit percona-cluster/0 "open-port 3306/tcp"

$ juju run --unit percona-cluster/0 "opened-ports --endpoints"
3306/tcp (*)
1337/udp (db-admin)

# Try to open port for a bogus endpoint
$ juju run --unit percona-cluster/0 "open-port 3306/tcp --endpoints bogus"

$ juju debug-log
...
ERROR juju.worker.uniter.context cannot apply changes: cannot open/close ports: open port range: endpoint "bogus" for unit "percona-cluster/0" not found
```

## Documentation changes

There is a card to do an audit of the related docs after all chunks of work land and make sure everything reflects the changes.
